src/probnum/filtsmooth/_gaussfiltsmooth/_kalman.py

```diff
@@ -231,121 +276,60 @@ def filtered_states_generator(
         --------
         RegressionProblem: a regression problem data class
         """
+
+        # Process inputs
         dataset, times = regression_problem.observations, regression_problem.locations
+        if not isinstance(measurement_model, abc.Iterable):
+            measurement_model = itertools.repeat(measurement_model, len(times))
 
-        _linearise_update_at = (
-            None if _previous_posterior is None else _previous_posterior(times[0])
-        )
+        # It is not clear at the moment how to implement this cleanly.
+        if not np.all(np.diff(times) > 0):
+            raise ValueError(
+                "Gaussian filtering cannot handle repeating time points currently."
+            )
 
-        info_dict = {"pred_rv": self.initrv, "info_pred": {}}
-        filtrv, info_dict["info_upd"] = self.update(
-            data=dataset[0],
-            rv=self.initrv,
-            time=times[0],
-            _linearise_at=_linearise_update_at,
-        )
-        yield filtrv, info_dict
+        # Initialise
+        t_old = self.prior_process.initarg
+        curr_rv = self.prior_process.initrv
 
-        for idx in range(1, len(times)):
-            _linearise_predict_at = (
-                None
-                if _previous_posterior is None
-                else _previous_posterior(times[idx - 1])
-            )
-            _linearise_update_at = (
-                None if _previous_posterior is None else _previous_posterior(times[idx])
+        # Iterate over data and measurement models
+        for t, data, measmod in itertools.zip_longest(
+            times, dataset, measurement_model, fillvalue=None
+        ):
+            if t is None or data is None or measmod is None:
+                errormsg = (
+                    "The lengths of the dataset, times and"
+                    "measurement models are inconsistent."
+                )
+
+                # At first sight, it seems desirable to raise a Warning instead of throwing an
+                # error. However, the remaining code will not work if one of the three
+                # (t, data, measmod) is not defined. Therefore, we throw a ValueError.
+                raise ValueError(errormsg)
+
+            dt = t - t_old
+            info_dict = {}
+
+            # Predict if there is a time-increment
+            if dt > 0:
+                linearise_predict_at = (
+                    None if _previous_posterior is None else _previous_posterior(t_old)
+                )
+                output = self.prior_process.transition.forward_rv(
+                    curr_rv, t, dt=dt, _linearise_at=linearise_predict_at
+                )
+                curr_rv, info_dict["predict_info"] = output
+
+            # Update (even if there is no increment)
+            linearise_update_at = (
+                None if _previous_posterior is None else _previous_posterior(t)
             )
```

Comment:
Are we still planning to get rid of the `linearise_at` arguments? Will this be dealt with in a separate PR?

Reply:
(Also regarding the performance, this might be an issue for the rather near than far future)

Reply:
yes, this should be changed soon. Should I do this in here, on shall we tackle it in a separate PR? I would tend towards the latter option, because it is a little bit orthogonal to the current PR

Reply:
Yep, let's handle that in a separate PR